src/wavm/mpi.cpp

```diff
@@ -843,6 +1202,202 @@ WAVM_DEFINE_INTRINSIC_FUNCTION(env,
     return MPI_SUCCESS;
 }
 
+/**
+ * Makes a new communicator to which Cartesian topology information has been
+ * attached.
+ *
+ * Reference implementation:
+ * https://github.com/open-mpi/ompi/blob/master/ompi/mca/topo/base/topo_base_cart_create.c
+ */
+WAVM_DEFINE_INTRINSIC_FUNCTION(env,
+                               "MPI_Cart_create",
+                               I32,
+                               MPI_Cart_create,
+                               I32 commOld,
+                               I32 ndims,
+                               I32 dims,
+                               I32 periods,
+                               I32 reorder,
+                               I32 newCommPtr)
+{
+    faabric::util::getLogger()->debug("S - MPI_Cart_create {} {} {} {} {} {}",
+                                      commOld,
+                                      ndims,
+                                      dims,
+                                      periods,
+                                      reorder,
+                                      newCommPtr);
+
+    ContextWrapper ctx(commOld);
+
+    // Set up the communicator object in the WASM memory. Note that we return
+    // a pointer to it.
+    I32 tmpPtr = Runtime::memoryRef<I32>(ctx.memory, newCommPtr);
+    faabric_communicator_t* newComm =
+      &Runtime::memoryRef<faabric_communicator_t>(ctx.memory, tmpPtr);
+
+    // Cast the original communicator
+    auto origComm =
+      Runtime::memoryRef<faabric_communicator_t>(ctx.memory, commOld);
+
+    // Copy the values from the old the the new
+    *newComm = origComm;
+
+    return MPI_SUCCESS;
+}
+
+/**
+ * Determines process rank in communicator given Cartesian location.
+ */
+WAVM_DEFINE_INTRINSIC_FUNCTION(env,
+                               "MPI_Cart_rank",
+                               I32,
+                               MPI_Cart_rank,
+                               I32 comm,
+                               I32 coords,
+                               I32 rankPtr)
+{
+    faabric::util::getLogger()->debug(
+      "S - MPI_Cart_rank {} {} {}", comm, coords, rankPtr);
+
+    ContextWrapper ctx(comm);
+
+    int* coordsArray = Runtime::memoryArrayPtr<int>(
+      ctx.memory, (Uptr)coords, MPI_CART_MAX_DIMENSIONS);
+    int sideLength =
+      static_cast<int>(std::floor(std::sqrt(ctx.world.getSize())));
+    int rank = sideLength * coordsArray[0] + coordsArray[1];
+    ctx.writeMpiResult<int>(rankPtr, rank);
```

Comment:
This logic should be moved into the world itself and have a small test. The function might be `ctx.world.getCartesianRank()`